// @tarojs/webpack-runner/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const opn = require("opn");
const path = require("path");
const url_1 = require("url");
const webpack = require("webpack");
const WebpackDevServer = require("webpack-dev-server");
const build_conf_1 = require("./config/build.conf");
const dev_conf_1 = require("./config/dev.conf");
const devServer_conf_1 = require("./config/devServer.conf");
const prod_conf_1 = require("./config/prod.conf");
const util_1 = require("./util");
const logHelper_1 = require("./util/logHelper");
const customizeChain = (chain, customizeFunc) => {
    if (customizeFunc instanceof Function) {
        customizeFunc(chain, webpack);
    }
};
const warnConfigWebpack = () => {
    console.log(chalk_1.default.yellow(`taro@1.2.18版本开始，h5.webpack配置项已经停止支持。作为代替，请使用配置项h5.webpackChain，文档：https://nervjs.github.io/taro/docs/config-detail.html#h5webpackchain`));
};
const warnConfigEnableDll = () => {
    console.log(chalk_1.default.yellow(`taro@1.2.18版本开始，taro在h5端加强了代码体积控制，抽离dll的收益已经微乎其微，同时也容易导致一些问题（#1800等）。所以h5.enableDll配置项暂时移除。`));
};
const buildProd = (appPath, config) => {
    return new Promise((resolve, reject) => {
        const webpackChain = prod_conf_1.default(appPath, config);
        customizeChain(webpackChain, config.webpackChain);
        if (config.webpack) {
            warnConfigWebpack();
        }
        const webpackConfig = webpackChain.toConfig();
        const compiler = webpack(webpackConfig);
        logHelper_1.bindProdLogger(compiler);
        compiler.run((err) => {
            if (err) {
                logHelper_1.printBuildError(err);
                return reject(err);
            }
            resolve();
        });
    });
};
const buildDev = (appPath, config) => __awaiter(this, void 0, void 0, function* () {
    return new Promise((resolve, reject) => {
        const conf = build_conf_1.default(config);
        const routerConfig = config.router || {};
        const routerMode = routerConfig.mode === 'browser' ? 'browser' : 'hash';
        const routerBasename = routerConfig.basename || '/';
        const publicPath = conf.publicPath ? util_1.addLeadingSlash(util_1.addTrailingSlash(conf.publicPath)) : '/';
        const outputPath = path.join(appPath, conf.outputRoot);
        const customDevServerOption = config.devServer || {};
        const webpackChain = dev_conf_1.default(appPath, config);
        customizeChain(webpackChain, config.webpackChain);
        if (config.webpack) {
            warnConfigWebpack();
        }
        const devServerOptions = util_1.recursiveMerge({
            publicPath,
            contentBase: outputPath,
            historyApiFallback: {
                index: publicPath
            }
        }, devServer_conf_1.default, customDevServerOption);
        const devUrl = url_1.format({
            protocol: devServerOptions.https ? 'https' : 'http',
            hostname: devServerOptions.host,
            port: devServerOptions.port,
            pathname: routerMode === 'browser' ? routerBasename : '/'
        });
        const webpackConfig = webpackChain.toConfig();
        WebpackDevServer.addDevServerEntrypoints(webpackConfig, devServerOptions);
        const compiler = webpack(webpackConfig);
        logHelper_1.bindDevLogger(devUrl, compiler);
        const server = new WebpackDevServer(compiler, devServerOptions);
        server.listen(devServerOptions.port, devServerOptions.disableHostCheck ? '0.0.0.0' : devServerOptions.host, err => {
            if (err) {
                reject();
                return console.log(err);
            }
            resolve();
            /* 补充处理devServer.open配置 */
            if (devServerOptions.open) {
                opn(devUrl);
            }
        });
    });
});
exports.default = (appPath, config) => __awaiter(this, void 0, void 0, function* () {
    if (config.isWatch) {
        try {
            yield buildDev(appPath, config);
        }
        catch (e) {
            console.error(e);
        }
    }
    else {
        if ('enableDll' in config) {
            warnConfigEnableDll();
        }
        try {
            yield buildProd(appPath, config);
        }
        catch (e) {
            console.error(e);
            process.exit(1);
        }
    }
});
