// @tarojs/router/dist/router.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('nervjs'), require('@tarojs/taro-h5')) :
  typeof define === 'function' && define.amd ? define(['exports', 'nervjs', '@tarojs/taro-h5'], factory) :
  (factory((global.Router = {}),global.Nerv,global.taroH5));
}(this, (function (exports,Nerv,taroH5) { 'use strict';

  Nerv = Nerv && Nerv.hasOwnProperty('default') ? Nerv['default'] : Nerv;

  function _typeof(obj) {
    if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
      _typeof = function (obj) {
        return typeof obj;
      };
    } else {
      _typeof = function (obj) {
        return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      };
    }

    return _typeof(obj);
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
  }

  function _get(object, property, receiver) {
    if (object === null) object = Function.prototype;
    var desc = Object.getOwnPropertyDescriptor(object, property);

    if (desc === undefined) {
      var parent = Object.getPrototypeOf(object);

      if (parent === null) {
        return undefined;
      } else {
        return _get(parent, property, receiver);
      }
    } else if ("value" in desc) {
      return desc.value;
    } else {
      var getter = desc.get;

      if (getter === undefined) {
        return undefined;
      }

      return getter.call(receiver);
    }
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function");
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  function _assertThisInitialized(self) {
    if (self === void 0) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return self;
  }

  function _possibleConstructorReturn(self, call) {
    if (call && (typeof call === "object" || typeof call === "function")) {
      return call;
    }

    return _assertThisInitialized(self);
  }

  /*!
   * EventEmitter v5.2.4 - git.io/ee
   * Unlicense - http://unlicense.org/
   * Oliver Caldwell - http://oli.me.uk/
   * @preserve
   */

  /**
   * Class for managing events.
   * Can be extended to provide event functionality in other classes.
   *
   * @class EventEmitter Manages event registering and emitting.
   */
  var EventEmitter = function EventEmitter() {}; // Shortcuts to improve speed and size


  var proto = EventEmitter.prototype;
  var originalGlobalValue = exports.EventEmitter;
  /**
   * Finds the index of the listener for the event in its storage array.
   *
   * @param {Function[]} listeners Array of listeners to search through.
   * @param {Function} listener Method to look for.
   * @return {Number} Index of the specified listener, -1 if not found
   * @api private
   */

  function indexOfListener(listeners, listener) {
    var i = listeners.length;

    while (i--) {
      if (listeners[i].listener === listener) {
        return i;
      }
    }

    return -1;
  }
  /**
   * Alias a method while keeping the context correct, to allow for overwriting of target method.
   *
   * @param {String} name The name of the target method.
   * @return {Function} The aliased method
   * @api private
   */


  function alias(name) {
    return function aliasClosure() {
      return this[name].apply(this, arguments);
    };
  }
  /**
   * Returns the listener array for the specified event.
   * Will initialise the event object and listener arrays if required.
   * Will return an object if you use a regex search. The object contains keys for each matched event. So /ba[rz]/ might return an object containing bar and baz. But only if you have either defined them with defineEvent or added some listeners to them.
   * Each property in the object response is an array of listener functions.
   *
   * @param {String|RegExp} evt Name of the event to return the listeners from.
   * @return {Function[]|Object} All listener functions for the event.
   */


  proto.getListeners = function getListeners(evt) {
    var events = this._getEvents();

    var response;
    var key; // Return a concatenated array of all matching events if
    // the selector is a regular expression.

    if (evt instanceof RegExp) {
      response = {};

      for (key in events) {
        if (events.hasOwnProperty(key) && evt.test(key)) {
          response[key] = events[key];
        }
      }
    } else {
      response = events[evt] || (events[evt] = []);
    }

    return response;
  };
  /**
   * Takes a list of listener objects and flattens it into a list of listener functions.
   *
   * @param {Object[]} listeners Raw listener objects.
   * @return {Function[]} Just the listener functions.
   */


  proto.flattenListeners = function flattenListeners(listeners) {
    var flatListeners = [];
    var i;

    for (i = 0; i < listeners.length; i += 1) {
      flatListeners.push(listeners[i].listener);
    }

    return flatListeners;
  };
  /**
   * Fetches the requested listeners via getListeners but will always return the results inside an object. This is mainly for internal use but others may find it useful.
   *
   * @param {String|RegExp} evt Name of the event to return the listeners from.
   * @return {Object} All listener functions for an event in an object.
   */


  proto.getListenersAsObject = function getListenersAsObject(evt) {
    var listeners = this.getListeners(evt);
    var response;

    if (listeners instanceof Array) {
      response = {};
      response[evt] = listeners;
    }

    return response || listeners;
  };

  function isValidListener(listener) {
    if (typeof listener === 'function' || listener instanceof RegExp) {
      return true;
    } else if (listener && _typeof(listener) === 'object') {
      return isValidListener(listener.listener);
    } else {
      return false;
    }
  }
  /**
   * Adds a listener function to the specified event.
   * The listener will not be added if it is a duplicate.
   * If the listener returns true then it will be removed after it is called.
   * If you pass a regular expression as the event name then the listener will be added to all events that match it.
   *
   * @param {String|RegExp} evt Name of the event to attach the listener to.
   * @param {Function} listener Method to be called when the event is emitted. If the function returns true then it will be removed after calling.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.addListener = function addListener(evt, listener) {
    if (!isValidListener(listener)) {
      throw new TypeError('listener must be a function');
    }

    var listeners = this.getListenersAsObject(evt);
    var listenerIsWrapped = _typeof(listener) === 'object';
    var key;

    for (key in listeners) {
      if (listeners.hasOwnProperty(key) && indexOfListener(listeners[key], listener) === -1) {
        listeners[key].push(listenerIsWrapped ? listener : {
          listener: listener,
          once: false
        });
      }
    }

    return this;
  };
  /**
   * Alias of addListener
   */


  proto.on = alias('addListener');
  /**
   * Semi-alias of addListener. It will add a listener that will be
   * automatically removed after its first execution.
   *
   * @param {String|RegExp} evt Name of the event to attach the listener to.
   * @param {Function} listener Method to be called when the event is emitted. If the function returns true then it will be removed after calling.
   * @return {Object} Current instance of EventEmitter for chaining.
   */

  proto.addOnceListener = function addOnceListener(evt, listener) {
    return this.addListener(evt, {
      listener: listener,
      once: true
    });
  };
  /**
   * Alias of addOnceListener.
   */


  proto.once = alias('addOnceListener');
  /**
   * Defines an event name. This is required if you want to use a regex to add a listener to multiple events at once. If you don't do this then how do you expect it to know what event to add to? Should it just add to every possible match for a regex? No. That is scary and bad.
   * You need to tell it what event names should be matched by a regex.
   *
   * @param {String} evt Name of the event to create.
   * @return {Object} Current instance of EventEmitter for chaining.
   */

  proto.defineEvent = function defineEvent(evt) {
    this.getListeners(evt);
    return this;
  };
  /**
   * Uses defineEvent to define multiple events.
   *
   * @param {String[]} evts An array of event names to define.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.defineEvents = function defineEvents(evts) {
    for (var i = 0; i < evts.length; i += 1) {
      this.defineEvent(evts[i]);
    }

    return this;
  };
  /**
   * Removes a listener function from the specified event.
   * When passed a regular expression as the event name, it will remove the listener from all events that match it.
   *
   * @param {String|RegExp} evt Name of the event to remove the listener from.
   * @param {Function} listener Method to remove from the event.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.removeListener = function removeListener(evt, listener) {
    var listeners = this.getListenersAsObject(evt);
    var index;
    var key;

    for (key in listeners) {
      if (listeners.hasOwnProperty(key)) {
        index = indexOfListener(listeners[key], listener);

        if (index !== -1) {
          listeners[key].splice(index, 1);
        }
      }
    }

    return this;
  };
  /**
   * Alias of removeListener
   */


  proto.off = alias('removeListener');
  /**
   * Adds listeners in bulk using the manipulateListeners method.
   * If you pass an object as the first argument you can add to multiple events at once. The object should contain key value pairs of events and listeners or listener arrays. You can also pass it an event name and an array of listeners to be added.
   * You can also pass it a regular expression to add the array of listeners to all events that match it.
   * Yeah, this function does quite a bit. That's probably a bad thing.
   *
   * @param {String|Object|RegExp} evt An event name if you will pass an array of listeners next. An object if you wish to add to multiple events at once.
   * @param {Function[]} [listeners] An optional array of listener functions to add.
   * @return {Object} Current instance of EventEmitter for chaining.
   */

  proto.addListeners = function addListeners(evt, listeners) {
    // Pass through to manipulateListeners
    return this.manipulateListeners(false, evt, listeners);
  };
  /**
   * Removes listeners in bulk using the manipulateListeners method.
   * If you pass an object as the first argument you can remove from multiple events at once. The object should contain key value pairs of events and listeners or listener arrays.
   * You can also pass it an event name and an array of listeners to be removed.
   * You can also pass it a regular expression to remove the listeners from all events that match it.
   *
   * @param {String|Object|RegExp} evt An event name if you will pass an array of listeners next. An object if you wish to remove from multiple events at once.
   * @param {Function[]} [listeners] An optional array of listener functions to remove.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.removeListeners = function removeListeners(evt, listeners) {
    // Pass through to manipulateListeners
    return this.manipulateListeners(true, evt, listeners);
  };
  /**
   * Edits listeners in bulk. The addListeners and removeListeners methods both use this to do their job. You should really use those instead, this is a little lower level.
   * The first argument will determine if the listeners are removed (true) or added (false).
   * If you pass an object as the second argument you can add/remove from multiple events at once. The object should contain key value pairs of events and listeners or listener arrays.
   * You can also pass it an event name and an array of listeners to be added/removed.
   * You can also pass it a regular expression to manipulate the listeners of all events that match it.
   *
   * @param {Boolean} remove True if you want to remove listeners, false if you want to add.
   * @param {String|Object|RegExp} evt An event name if you will pass an array of listeners next. An object if you wish to add/remove from multiple events at once.
   * @param {Function[]} [listeners] An optional array of listener functions to add/remove.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.manipulateListeners = function manipulateListeners(remove, evt, listeners) {
    var i;
    var value;
    var single = remove ? this.removeListener : this.addListener;
    var multiple = remove ? this.removeListeners : this.addListeners; // If evt is an object then pass each of its properties to this method

    if (_typeof(evt) === 'object' && !(evt instanceof RegExp)) {
      for (i in evt) {
        if (evt.hasOwnProperty(i) && (value = evt[i])) {
          // Pass the single listener straight through to the singular method
          if (typeof value === 'function') {
            single.call(this, i, value);
          } else {
            // Otherwise pass back to the multiple function
            multiple.call(this, i, value);
          }
        }
      }
    } else {
      // So evt must be a string
      // And listeners must be an array of listeners
      // Loop over it and pass each one to the multiple method
      i = listeners.length;

      while (i--) {
        single.call(this, evt, listeners[i]);
      }
    }

    return this;
  };
  /**
   * Removes all listeners from a specified event.
   * If you do not specify an event then all listeners will be removed.
   * That means every event will be emptied.
   * You can also pass a regex to remove all events that match it.
   *
   * @param {String|RegExp} [evt] Optional name of the event to remove all listeners for. Will remove from every event if not passed.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.removeEvent = function removeEvent(evt) {
    var type = _typeof(evt);

    var events = this._getEvents();

    var key; // Remove different things depending on the state of evt

    if (type === 'string') {
      // Remove all listeners for the specified event
      delete events[evt];
    } else if (evt instanceof RegExp) {
      // Remove all events matching the regex.
      for (key in events) {
        if (events.hasOwnProperty(key) && evt.test(key)) {
          delete events[key];
        }
      }
    } else {
      // Remove all listeners in all events
      delete this._events;
    }

    return this;
  };
  /**
   * Alias of removeEvent.
   *
   * Added to mirror the node API.
   */


  proto.removeAllListeners = alias('removeEvent');
  /**
   * Emits an event of your choice.
   * When emitted, every listener attached to that event will be executed.
   * If you pass the optional argument array then those arguments will be passed to every listener upon execution.
   * Because it uses `apply`, your array of arguments will be passed as if you wrote them out separately.
   * So they will not arrive within the array on the other side, they will be separate.
   * You can also pass a regular expression to emit to all events that match it.
   *
   * @param {String|RegExp} evt Name of the event to emit and execute listeners for.
   * @param {Array} [args] Optional array of arguments to be passed to each listener.
   * @return {Object} Current instance of EventEmitter for chaining.
   */

  proto.emitEvent = function emitEvent(evt, args) {
    var listenersMap = this.getListenersAsObject(evt);
    var listeners;
    var listener;
    var i;
    var key;
    var response;

    for (key in listenersMap) {
      if (listenersMap.hasOwnProperty(key)) {
        listeners = listenersMap[key].slice(0);

        for (i = 0; i < listeners.length; i++) {
          // If the listener returns true then it shall be removed from the event
          // The function is executed either with a basic call or an apply if there is an args array
          listener = listeners[i];

          if (listener.once === true) {
            this.removeListener(evt, listener.listener);
          }

          response = listener.listener.apply(this, args || []);

          if (response === this._getOnceReturnValue()) {
            this.removeListener(evt, listener.listener);
          }
        }
      }
    }

    return this;
  };
  /**
   * Alias of emitEvent
   */


  proto.trigger = alias('emitEvent');
  /**
   * Subtly different from emitEvent in that it will pass its arguments on to the listeners, as opposed to taking a single array of arguments to pass on.
   * As with emitEvent, you can pass a regex in place of the event name to emit to all events that match it.
   *
   * @param {String|RegExp} evt Name of the event to emit and execute listeners for.
   * @param {...*} Optional additional arguments to be passed to each listener.
   * @return {Object} Current instance of EventEmitter for chaining.
   */

  proto.emit = function emit(evt) {
    var args = Array.prototype.slice.call(arguments, 1);
    return this.emitEvent(evt, args);
  };
  /**
   * Sets the current value to check against when executing listeners. If a
   * listeners return value matches the one set here then it will be removed
   * after execution. This value defaults to true.
   *
   * @param {*} value The new value to check for when executing listeners.
   * @return {Object} Current instance of EventEmitter for chaining.
   */


  proto.setOnceReturnValue = function setOnceReturnValue(value) {
    this._onceReturnValue = value;
    return this;
  };
  /**
   * Fetches the current value to check against when executing listeners. If
   * the listeners return value matches this one then it should be removed
   * automatically. It will return true by default.
   *
   * @return {*|Boolean} The current value to check for or the default, true.
   * @api private
   */


  proto._getOnceReturnValue = function _getOnceReturnValue() {
    if (this.hasOwnProperty('_onceReturnValue')) {
      return this._onceReturnValue;
    } else {
      return true;
    }
  };
  /**
   * Fetches the events object and creates one if required.
   *
   * @return {Object} The events storage object.
   * @api private
   */


  proto._getEvents = function _getEvents() {
    return this._events || (this._events = {});
  };
  /**
   * Reverts the global {@link EventEmitter} to its previous value and returns a reference to this version.
   *
   * @return {Function} Non conflicting EventEmitter class.
   */


  EventEmitter.noConflict = function noConflict() {
    exports.EventEmitter = originalGlobalValue;
    return EventEmitter;
  };

  /**
   * Filename: /src/lib/utils.js
   * Created Date: 2018-05-27 03:48:58
   * Author: Littly
   * Copyright (c) 2018 JD.COM
   */

  /**
   * 将hash路径转换为全路径
   *
   * @param {string} url hash路径
   * @returns {string} 带hash的全路径
   */
  var convertToFullUrl = function convertToFullUrl(url) {
    var hashIndex = window.location.href.indexOf('#');
    return window.location.href.slice(0, hashIndex >= 0 ? hashIndex : 0) + '#' + url;
  };
  /**
   * 将location.hash更换为新的hash
   *
   * @param {string} url 待跳转的hash路径
   */


  var pushHash = function pushHash(url) {
    window.location.hash = url;
  };
  /**
   * 将当前的hash路径替换为新的
   *
   * @param {string} param0
   * @param {string} param0.url
   * @param {string} param0.state
   */


  var replaceHash = function replaceHash(_ref) {
    var url = _ref.url,
        state = _ref.state;
    window.history.replaceState(state, '', convertToFullUrl(url));
  };
  /**
   * 获取当前hash路径
   *
   * @returns {string} hash路径
   */


  var getCurrentHash = function getCurrentHash() {
    var href = window.location.href;
    var hashIndex = href.indexOf('#');
    return hashIndex === -1 ? '' : href.substring(hashIndex + 1);
  };
  /**
   * 标准化路径（绝对路径，不带问号）
   *
   * @param {string} url 待转换的路径
   */


  var normalizeUrl = function normalizeUrl(url) {
    url = url.replace(/\?$/, '');
    return url.charAt(0) === '/' ? url : "/".concat(url);
  };

  var counter = 0;
  /**
   * @typedef {object} location
   * @property {string} routerIdx 页面对应的routerIdx
   * @property {string} fullUrl 路径
   * @property {string} pageId  location的唯一标识
   * @property {string} params query中的参数
   * @property {string} url 路径的主体部分
   */

  /**
   * 生成location对象
   *
   * @param {string} fullUrl 待处理的路径
   * @param {number} state 页面对应的state
   * @return {location} location对象
   */

  var createLocation = function createLocation(fullUrl, state) {
    fullUrl = decodeURIComponent(fullUrl);
    var url = fullUrl.split('?')[0];
    var params = {};
    var queryIndex = fullUrl.indexOf('?');

    if (queryIndex !== -1) {
      var queryString = fullUrl.substring(queryIndex + 1);
      queryString.split('&').forEach(function (pair) {
        var temp = pair.split('=');
        params[temp[0]] = temp[1];
      });
    }

    return {
      state: state,
      fullUrl: fullUrl,
      pageId: counter++,
      params: params,
      url: url
    };
  };

  function isAbsolute(pathname) {
    return pathname.charAt(0) === '/';
  } // About 1.5x faster than the two-arg version of Array#splice()


  function spliceOne(list, index) {
    for (var i = index, k = i + 1, n = list.length; k < n; i += 1, k += 1) {
      list[i] = list[k];
    }

    list.pop();
  } // This implementation is based heavily on node's url.parse


  function resolvePathname(to) {
    var from = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
    var toParts = to && to.split('/') || [];
    var fromParts = from && from.split('/') || [];
    var isToAbs = to && isAbsolute(to);
    var isFromAbs = from && isAbsolute(from);
    var mustEndAbs = isToAbs || isFromAbs;

    if (to && isAbsolute(to)) {
      // to is absolute
      fromParts = toParts;
    } else if (toParts.length) {
      // to is relative, drop the filename
      fromParts.pop();
      fromParts = fromParts.concat(toParts);
    }

    if (!fromParts.length) {
      return '/';
    }

    var hasTrailingSlash;

    if (fromParts.length) {
      var last = fromParts[fromParts.length - 1];
      hasTrailingSlash = last === '.' || last === '..' || last === '';
    } else {
      hasTrailingSlash = false;
    }

    var up = 0;

    for (var i = fromParts.length; i >= 0; i--) {
      var part = fromParts[i];

      if (part === '.') {
        spliceOne(fromParts, i);
      } else if (part === '..') {
        spliceOne(fromParts, i);
        up++;
      } else if (up) {
        spliceOne(fromParts, i);
        up--;
      }
    }

    if (!mustEndAbs) {
      for (; up--; up) {
        fromParts.unshift('..');
      }
    }

    if (mustEndAbs && fromParts[0] !== '' && (!fromParts[0] || !isAbsolute(fromParts[0]))) {
      fromParts.unshift('');
    }

    var result = fromParts.join('/');

    if (hasTrailingSlash && result.substr(-1) !== '/') {
      result += '/';
    }

    return result;
  }

  var navigateTo = function navigateTo() {
    var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var current = h.now();
    var currentUrl = current.url;
    opts.url = resolvePathname(opts.url, currentUrl);
    h.push(opts);
  };

  var navigateBack = function navigateBack() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      delta: 1
    },
        delta = _ref.delta;

    window.history.go(-1 * delta);
  };

  var redirectTo = function redirectTo() {
    var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var success = opts.success;
    var fail = opts.fail;
    var complete = opts.complete;
    h.replace({
      url: opts.url,
      success: success,
      fail: fail,
      complete: complete
    });
  };

  var history = window.history;
  var eventEmitter = new EventEmitter();
  var historyState = parseInt(history.state, 10) || 0;
  var initUrl = normalizeUrl(getCurrentHash());
  var initLocation = createLocation(initUrl, historyState);
  var ignoredUrl = null;
  var HISTORYKEY = 'taroRouterHistory';

  var generateSerializer = function generateSerializer(obj) {
    return function () {
      localStorage.setItem(HISTORYKEY, JSON.stringify(obj));
    };
  };

  var History =
  /*#__PURE__*/
  function () {
    function History() {
      var _this = this;

      _classCallCheck(this, History);

      Object.defineProperty(this, "onHashchange", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(e) {
          history.replaceState(historyState, '', '');
        }
      });
      Object.defineProperty(this, "onPopstate", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(e) {
          var nextUrl = normalizeUrl(getCurrentHash());

          if (typeof e.state !== 'number') {
            if (nextUrl === ignoredUrl) return;
            ignoredUrl = null;
            navigateTo({
              url: nextUrl
            });
            return;
          }

          var isBackPage = e.state < historyState;

          if (isBackPage) {
            _this.goBack({
              url: nextUrl,
              state: e.state
            });
          } else {
            navigateTo({
              url: nextUrl,
              state: e.state,
              isForward: true
            });
          }
        }
      });
      Object.defineProperty(this, "listen", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(fn) {
          eventEmitter.on('change', fn);
        }
      });
      Object.defineProperty(this, "unlisten", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(fn) {
          eventEmitter.off('change', fn);
        }
      });
      Object.defineProperty(this, "emit", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(location, type, opts) {
          eventEmitter.emit('change', location, type, opts);
        }
      });
      Object.defineProperty(this, "push", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(_ref) {
          var url = _ref.url,
              isForward = _ref.isForward,
              state = _ref.state;
          var nextUrl = normalizeUrl(url);

          var currentUrl = _this.now().url;

          if (!isForward && nextUrl === currentUrl) return;
          /**
           * 带state 是forward 否则是navigate
           */

          if (!state) {
            historyState += 1;
          } else {
            historyState = state;
          }

          var location = createLocation(nextUrl, historyState);

          _this.locationStack.push(location);

          _this.serializeStack();

          _this.emit(location, 'PUSH');

          ignoredUrl = nextUrl;

          if (!isForward) {
            pushHash(nextUrl);
          }
        }
      });
      Object.defineProperty(this, "replace", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(_ref2) {
          var url = _ref2.url;
          var nextUrl = normalizeUrl(url);

          var currentUrl = _this.now().url;

          if (nextUrl === currentUrl) return;
          var location = createLocation(nextUrl, historyState);

          _this.locationStack.pop();

          _this.locationStack.push(location);

          _this.serializeStack();

          _this.emit(location, 'REPLACE');

          ignoredUrl = nextUrl;
          replaceHash({
            url: nextUrl,
            state: historyState
          });
          history.replaceState;
        }
      });
      Object.defineProperty(this, "goBack", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(_ref3) {
          var url = _ref3.url,
              state = _ref3.state;

          var len = _this.len();

          var foundStateIdx = _this.locationStack.findIndex(function (v) {
            return v.state === state;
          });

          var delta = foundStateIdx > -1 ? _this.locationStack.length - foundStateIdx - 1 : 1;

          if (len > delta) {
            _this.locationStack.splice(-delta);

            _this.serializeStack();

            var location = _this.now();

            historyState = location.state;

            _this.emit(location, 'BACK', {
              delta: delta
            });
          } else if (delta <= 1 && url) {
            var _location = createLocation(normalizeUrl(url), state);

            historyState = state;
            _this.locationStack.length = 1;
            _this.locationStack[0] = _location;

            _this.serializeStack();

            _this.emit(_location, 'BACK', {
              delta: delta
            });
          } else {
            return console.warn('goBack delta out of range');
          }
        }
      });
      history.replaceState(historyState, '', null);
      window.addEventListener('hashchange', this.onHashchange);
      window.addEventListener('popstate', this.onPopstate);
      var stack = this.deserialize();

      if (this.validateStack(stack)) {
        this.locationStack = stack;
        this.serializeStack = generateSerializer(this.locationStack);
      } else {
        console.warn('Stack in storage invalid');
        this.locationStack = [initLocation];
        this.serializeStack = generateSerializer(this.locationStack);
        this.serializeStack();
      }
    }

    _createClass(History, [{
      key: "validateStack",
      value: function validateStack(stack) {
        if (stack.length === 0) return false;
        var historyMap = {};
        var hasDuplicateState = stack.some(function (v) {
          var state = v.state;
          if (historyMap[state]) return true;
          historyMap[state] = true;
        });
        if (hasDuplicateState) return false;
        var stackCurrent = stack[stack.length - 1];
        var isThisone = stackCurrent.state === initLocation.state && stackCurrent.fullUrl === initLocation.fullUrl;
        if (isThisone) return true;else return false;
      }
    }, {
      key: "deserialize",
      value: function deserialize() {
        try {
          var stackStr = localStorage.getItem(HISTORYKEY);
          var stack = JSON.parse(stackStr);
          if (!Array.isArray(stack)) return [];
          return stack;
        } catch (e) {
          return [];
        }
      }
    }, {
      key: "now",

      /**
       * 返回当前的location
       *
       * @returns {location} location对象
       */
      value: function now() {
        var current = this.locationStack.length - 1;
        return this.locationStack[current];
      }
      /**
       * 返回当前的history栈长度
       *
       * @returns {number} history.stack的长度
       */

    }, {
      key: "len",
      value: function len() {
        return this.locationStack.length;
      }
      /**
       * 在history.stack中push新的location
       *
       * @param {object} param0 push的对象
       */

    }]);

    return History;
  }();

  var h = new History();

  function styleInject(css, ref) {
    if (ref === void 0) ref = {};
    var insertAt = ref.insertAt;

    if (!css || typeof document === 'undefined') {
      return;
    }

    var head = document.head || document.getElementsByTagName('head')[0];
    var style = document.createElement('style');
    style.type = 'text/css';

    if (insertAt === 'top') {
      if (head.firstChild) {
        head.insertBefore(style, head.firstChild);
      } else {
        head.appendChild(style);
      }
    } else {
      head.appendChild(style);
    }

    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }
  }

  var css = ".taro_router {\r\n  position: absolute;\r\n  width: 100%;\r\n  height: 100%;\r\n}\r\n.taro_page {\r\n  position: absolute;\r\n  width: 100%;\r\n  top: 0;\r\n  left: 0;\r\n  background: #fff;\r\n  display: block;\r\n}\r\n\r\n.taro_page_show {\r\n  display: block;\r\n}\r\n\r\n.taro_page_hide {\r\n  display: none;\r\n}\r\n";
  styleInject(css);

  var registeredPages = {};
  var registeredPagesArr = [];

  var hasPage = function hasPage(pathname) {
    var isNormalPage = registeredPagesArr.some(function (v) {
      return pathname === v[0];
    });
    var isIndex = /^\/(index(\.html)?)?[^/]*$/.test(pathname);
    if (isNormalPage || isIndex) return true;
    return false;
  };

  var getPage = function getPage(pathname) {
    if (pathname in registeredPages) {
      return registeredPages[pathname];
    } else {
      return registeredPagesArr[0][1];
    }
  };

  var PAGESTATUS = {
    HIDDEN: 1,
    SHOWING: 2
  };

  var getWrappedComponent = function getWrappedComponent(component) {
    var Wrapped =
    /*#__PURE__*/
    function (_component) {
      _inherits(Wrapped, _component);

      function Wrapped(props, context) {
        var _this;

        _classCallCheck(this, Wrapped);

        _this = _possibleConstructorReturn(this, (Wrapped.__proto__ || Object.getPrototypeOf(Wrapped)).call(this, props, context));
        Object.defineProperty(_assertThisInitialized(_this), "__pageStatus", {
          configurable: true,
          enumerable: true,
          writable: true,
          value: PAGESTATUS.SHOWING
        });
        _this.locationState = props._$router.state;
        _this.$router = props._$router;
        return _this;
      }

      _createClass(Wrapped, [{
        key: "componentWillReceiveProps",
        value: function componentWillReceiveProps(nextProps) {
          _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentWillReceiveProps", this) && _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentWillReceiveProps", this).call(this, nextProps);
          var nextLocation = nextProps._$router;
          var lastShouldShow = this.props._$router.state === this.locationState;
          var nextShouldShow = nextLocation.state === this.locationState;
          if (lastShouldShow === nextShouldShow) return;
          this.$router = nextLocation;

          if (nextShouldShow) {
            this.__pageStatus = PAGESTATUS.SHOWING;
            this.forceUpdate();
            this.componentDidShow && this.componentDidShow();
          } else {
            this.__pageStatus = PAGESTATUS.HIDDEN;
            this.forceUpdate();
            this.componentDidHide && this.componentDidHide();
          }
        }
      }, {
        key: "componentDidMount",
        value: function componentDidMount() {
          _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidMount", this) && _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidMount", this).call(this);
          this.defaultShow = true;
          _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidShow", this) && _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidShow", this).call(this);
        }
      }, {
        key: "componentWillUnmount",
        value: function componentWillUnmount() {
          _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidHide", this) && _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentDidHide", this).call(this);
          _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentWillUnmount", this) && _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "componentWillUnmount", this).call(this);
        }
      }, {
        key: "render",
        value: function render() {
          var containerStyle = this.__pageStatus === PAGESTATUS.SHOWING ? {
            display: 'block'
          } : {
            display: 'none'
          };
          return Nerv.createElement("div", {
            className: "taro_page",
            dataState: this.locationState,
            style: containerStyle
          }, _get(Wrapped.prototype.__proto__ || Object.getPrototypeOf(Wrapped.prototype), "render", this).call(this));
        }
      }]);

      return Wrapped;
    }(component);

    return Wrapped;
  };
  /**
   * Router组件内保存了每次路由变化后渲染的组件
   */


  var Router =
  /*#__PURE__*/
  function (_Component) {
    _inherits(Router, _Component);

    function Router() {
      var _ref;

      var _temp, _this2;

      _classCallCheck(this, Router);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _possibleConstructorReturn(_this2, (_temp = _this2 = _possibleConstructorReturn(this, (_ref = Router.__proto__ || Object.getPrototypeOf(Router)).call.apply(_ref, [this].concat(args))), Object.defineProperty(_assertThisInitialized(_this2), "navigate", {
        configurable: true,
        enumerable: true,
        writable: true,
        value: function value(location, action) {
          var payload = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
          var fail = payload.fail,
              complete = payload.complete,
              success = payload.success,
              delta = payload.delta;
          var pathname;

          if (action === 'BACK') {
            if (_this2.constructor.pageStack.length <= delta) pathname = location.url;else return _this2.commit('BACK', null, {
              delta: delta
            });
          } else if (action === 'PUSH' || action === 'REPLACE') {
            pathname = location.url;
          }
          /* loadPage */


          if (!hasPage(pathname)) {
            fail && fail();
            complete && complete();
            return console.warn('page not found');
          }

          getPage(pathname)().then(function (loaded) {
            var comp = getWrappedComponent(loaded.default); // this.commit(action, Nerv.createElement(comp, {
            //   _$router: location
            // }), payload)

            _this2.commit(action, comp, payload);

            success && success();
          }).catch(function (e) {
            console.error(e);
            fail && fail();
          }).then(function () {
            complete && complete();
          });
        }
      }), _temp));
    }

    _createClass(Router, [{
      key: "commit",
      value: function commit(action, el, payload) {
        var delta = payload.delta;

        switch (action) {
          case 'PUSH':
            this.constructor.pageStack.push(el);
            this.forceUpdate();
            break;

          case 'REPLACE':
            var pageStack = this.constructor.pageStack;
            pageStack.pop();
            pageStack.push(el);
            this.forceUpdate();
            break;

          case 'BACK':
            if (el) {
              this.constructor.pageStack = [el];
              this.forceUpdate();
            } else {
              var _pageStack = this.constructor.pageStack;

              _pageStack.splice(-delta);

              this.forceUpdate();
            }

            break;

          default:
            console.warn('wrong action');
        }
      }
    }, {
      key: "getPages",
      value: function getPages() {
        var $router = h.now();
        this.constructor.pageInstStack = this.constructor.pageStack.map(function (Comp) {
          return Nerv.createElement(Comp, {
            _$router: $router
          });
        });
        return this.constructor.pageInstStack;
      }
    }, {
      key: "componentDidMount",
      value: function componentDidMount() {
        h.listen(this.navigate);
        this.navigate(h.now(), 'PUSH');
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        h.unlisten(this.navigate);
      }
    }, {
      key: "render",
      value: function render() {
        return Nerv.createElement("div", {
          className: "taro_router"
        }, this.getPages());
      }
    }]);

    return Router;
  }(taroH5.Component);

  Object.defineProperty(Router, "pageStack", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: []
  });
  Object.defineProperty(Router, "pageInstStack", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: []
  });

  var getCurrentPages = function getCurrentPages(opts) {
    return Router.pageStack;
  };

  var initRouter = function initRouter(pageArr, taro) {
    registeredPagesArr = pageArr;
    pageArr.forEach(function (v) {
      var pageName = v[0];
      var pageLoader = v[1];
      registeredPages[pageName] = pageLoader;
    });
    taro.navigateTo = navigateTo;
    taro.navigateBack = navigateBack;
    taro.redirectTo = redirectTo;
    taro.getCurrentPages = getCurrentPages;
  };

  var index = {
    Router: Router,
    initRouter: initRouter
  };

  exports.default = index;
  exports.Router = Router;
  exports.initRouter = initRouter;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
